pcsx2/Ps2Float.cpp

```diff
@@ -0,0 +1,791 @@
+
+	if (result.Mantissa > 0)
+	{
+		int32_t leadingBitPosition = Ps2Float::GetMostSignificantBitPosition(result.Mantissa);
+
+		while (leadingBitPosition != IMPLICIT_LEADING_BIT_POS)
+		{
+			if (leadingBitPosition > IMPLICIT_LEADING_BIT_POS)
+			{
+				result.Mantissa >>= 1;
+
+				int32_t exp = ((int32_t)result.Exponent + 1);
+
+				if (exp > 255)
+					return result.Sign ? Min() : Max();
+
+				result.Exponent = (uint8_t)exp;
+
+				leadingBitPosition--;
+			}
+			else if (leadingBitPosition < IMPLICIT_LEADING_BIT_POS)
+			{
+				result.Mantissa <<= 1;
+
+				int32_t exp = ((int32_t)result.Exponent - 1);
+
+				if (exp <= 0)
+					return Ps2Float(result.Sign, 0, 0);
+
+				result.Exponent = (uint8_t)exp;
+
+				leadingBitPosition++;
+			}
+		}
+	}
+
+	result.Mantissa &= 0x7FFFFF;
+	return result.RoundTowardsZero();
+}
+
+Ps2Float Ps2Float::SolveAbnormalAdditionOrSubtractionOperation(Ps2Float a, Ps2Float b, bool add)
+{
+	uint32_t aval = a.AsUInt32();
+	uint32_t bval = b.AsUInt32();
+
+	if (aval == MAX_FLOATING_POINT_VALUE && bval == MAX_FLOATING_POINT_VALUE)
+		return add ? Max() : Ps2Float(0);
+
+	if (aval == MIN_FLOATING_POINT_VALUE && bval == MIN_FLOATING_POINT_VALUE)
+		return add ? Min() : Ps2Float(0);
+
+	if (aval == MIN_FLOATING_POINT_VALUE && bval == MAX_FLOATING_POINT_VALUE)
+		return add ? Ps2Float(0) : Min();
+
+	if (aval == MAX_FLOATING_POINT_VALUE && bval == MIN_FLOATING_POINT_VALUE)
+		return add ? Ps2Float(0) : Max();
+
+	if (aval == POSITIVE_INFINITY_VALUE && bval == POSITIVE_INFINITY_VALUE)
+		return add ? Max() : Ps2Float(0);
+
+	if (aval == NEGATIVE_INFINITY_VALUE && bval == POSITIVE_INFINITY_VALUE)
+		return add ? Ps2Float(0) : Min();
+
+	if (aval == POSITIVE_INFINITY_VALUE && bval == NEGATIVE_INFINITY_VALUE)
+		return add ? Ps2Float(0) : Max();
+
+	if (aval == NEGATIVE_INFINITY_VALUE && bval == NEGATIVE_INFINITY_VALUE)
+		return add ? Min() : Ps2Float(0);
+
+	if (aval == MAX_FLOATING_POINT_VALUE && bval == POSITIVE_INFINITY_VALUE)
+		return add ? Max() : Ps2Float(0x7F7FFFFE);
+
+	if (aval == MAX_FLOATING_POINT_VALUE && bval == NEGATIVE_INFINITY_VALUE)
+		return add ? Ps2Float(0x7F7FFFFE) : Max();
+
+	if (aval == MIN_FLOATING_POINT_VALUE && bval == POSITIVE_INFINITY_VALUE)
+		return add ? Ps2Float(0xFF7FFFFE) : Min();
+
+	if (aval == MIN_FLOATING_POINT_VALUE && bval == NEGATIVE_INFINITY_VALUE)
+		return add ? Min() : Ps2Float(0xFF7FFFFE);
+
+	if (aval == POSITIVE_INFINITY_VALUE && bval == MAX_FLOATING_POINT_VALUE)
+		return add ? Max() : Ps2Float(0xFF7FFFFE);
+
+	if (aval == POSITIVE_INFINITY_VALUE && bval == MIN_FLOATING_POINT_VALUE)
+		return add ? Ps2Float(0xFF7FFFFE) : Max();
+
+	if (aval == NEGATIVE_INFINITY_VALUE && bval == MAX_FLOATING_POINT_VALUE)
+		return add ? Ps2Float(0x7F7FFFFE) : Min();
+
+	if (aval == NEGATIVE_INFINITY_VALUE && bval == MIN_FLOATING_POINT_VALUE)
+		return add ? Min() : Ps2Float(0x7F7FFFFE);
+
+	Console.Error("Unhandled abnormal add/sub floating point operation");
+}
```

Comment:
You don't return anything here despite your function signature. It might be an abnormal case but you still need to make it deterministic (this is currently UB).
Should address MSVC warning -Wreturn-type